src/asm/format.c

```diff
@@ -0,0 +1,272 @@
+/*
+ * This file is part of RGBDS.
+ *
+ * Copyright (c) 2020, RGBDS contributors.
+ *
+ * SPDX-License-Identifier: MIT
+ */
+
+#include <inttypes.h>
+#include <math.h>
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "asm/format.h"
+#include "asm/warning.h"
+
+void fmt_UseCharacter(struct FormatSpec *fmt, int c)
+{
+	if (fmt->state == FORMAT_INVALID)
+		return;
+
+	switch (c) {
+	/* sign */
+	case ' ':
+	case '+':
+		if (fmt->state > FORMAT_SIGN)
+			goto invalid;
+		fmt->state = FORMAT_PREFIX;
+		fmt->sign = c;
+		break;
+
+	/* prefix */
+	case '#':
+		if (fmt->state > FORMAT_PREFIX)
+			goto invalid;
+		fmt->state = FORMAT_ALIGN;
+		fmt->prefix = true;
+		break;
+
+	/* align */
+	case '-':
+		if (fmt->state > FORMAT_ALIGN)
+			goto invalid;
+		fmt->state = FORMAT_WIDTH;
+		fmt->alignLeft = true;
+		break;
+
+	/* pad and width */
+	case '0':
+		if (fmt->state < FORMAT_WIDTH)
+			fmt->padZero = true;
+		/* fallthrough */
+	case '1':
+	case '2':
+	case '3':
+	case '4':
+	case '5':
+	case '6':
+	case '7':
+	case '8':
+	case '9':
+		if (fmt->state < FORMAT_WIDTH) {
+			fmt->state = FORMAT_WIDTH;
+			fmt->width = c - '0';
+		} else if (fmt->state == FORMAT_WIDTH) {
+			fmt->width = fmt->width * 10 + (c - '0');
+		} else if (fmt->state == FORMAT_FRAC) {
+			fmt->fracWidth = fmt->fracWidth * 10 + (c - '0');
+		} else {
+			goto invalid;
+		}
+		break;
+
+	case '.':
+		if (fmt->state > FORMAT_WIDTH)
+			goto invalid;
+		fmt->state = FORMAT_FRAC;
+		fmt->hasFrac = true;
+		break;
+
+	/* type */
+	case 'd':
+	case 'u':
+	case 'X':
+	case 'x':
+	case 'b':
+	case 'o':
+	case 'f':
+	case 's':
+		if (fmt->state >= FORMAT_DONE)
+			goto invalid;
+		fmt->state = FORMAT_DONE;
+		fmt->valid = true;
+		fmt->type = c;
+		break;
+
+invalid:
+	default:
+		fmt->state = FORMAT_INVALID;
+		fmt->valid = false;
```

Comment:
Doesn't that make `fmt->valid` equivalent to `FORMAT_INVALID`?

Reply:
No, a format starts out with `valid == false` and `state == FORMAT_SIGN` (no characters read yet, expecting the `<sign>` piece or further). When the whole format is read and reaches `state = FORMAT_DONE`, it also gets `valid = true`. But then if it reads another character (e.g. `{%02dspam:NUM}`) it reaches `state = FORMAT_INVALID` and `valid = false` and stays there.

Reply:
So then `valid` would be equivalent to `state == FORMAT_DONE`?

Reply:
After `readInterpolation` reads the format it checks for `!fmt.valid && fmt.state`, because if no format was given at all (e.g. `{NUM}`) then it would have `valid == false` and `state == 0` (`FORMAT_SIGN`). Later, before calling `fmt_PrintString` or `fmt_PrintNumber`, it checks for `!fmt.state` and sets a default format if no format was given.

When `STRFMT` is implemented, part or all of this logic may be encapsulated in `fmt_*` functions if it's reusable. (It mostly won't be, since `%` specifiers cannot be empty.)

Reply:
Could the logic be implemented as
```
if (fmt_Nothing()) { // state == 0
    // Default
} else if (fmt_Bad()) { // state != DONE
    // Bad user! Bad!
}
```

Reply:
I can wrap the state/valid checks in `fmt_` functions, but I think the current placement makes sense -- check and warn of an invalid format as soon as the format is read, before reading the whole interpolation string; and assign default formats for an unspecified format only after the string has been read and we know what type the symbol is.

Reply:
Would that be preferable? `fmt_IsEmpty(fmt) -> !fmt_state` and `fmt_IsInvalid(fmt) -> !fmt.valid && fmt.state`?

Reply:
Given the headaches that the symbol system rewrite has given me for similar reasons, I would strongly encourage such wrappers, if only to provide insight in what the checks are *meant* to check.

Reply:
Done.